### ISSUE TITLE
Can Datastore sync every offline changes of the same record to the server?

### ISSUE THREAD

**Issue by nubpro** [aws-amplify/amplify-js#6203]
**Is your feature request related to a problem? Please describe.**
In our use case, we would like keep track all the changes performed by the client during offline mode and sync all the changes back to the server when its back online.

1. Client is online and receives record A
2. Client goes offline
3. Client updates record A to _order accepted_ (offline)
4. Client updates record A to _order is being cooked_ (offline)
5. Client updates record A to _order is ready to be served_  (offline)
6. Client goes back online.

Basically, we would like to know all the changes that happened between steps 2 and 6, so we can keep them in a separate table for tracing purposes. These logs are very critical to our business requirement.

**Describe the solution you'd like**
Sync every OFFLINE changes of the same record to the server.

**Comment by wei** [comment]
Perhaps you can use an array for tracing logs in this case?

1. Client is online and receives record A
    
    ```
    {
       "id": "xxx",
       "statusLog": [],
       "lastStatus": null
    }
    ```

2. Client goes offline
3. Client updates record A to order accepted (offline)
    
    ```
    {
       "id": "xxx",
       "statusLog": [
           "accepted"
       ],
       "lastStatus": "accepted"
    }
    ```

4. Client updates record A to order is being cooked (offline)
    
    ```
    {
       "id": "xxx",
       "statusLog": [
           "accepted",
           "being cooked"
       ],
       "lastStatus": "being cooked"
    }
    ```

5. Client updates record A to order is ready to be served (offline)
    
    ```
    {
       "id": "xxx",
       "statusLog": [
           "accepted",
           "being cooked",
           "served"
       ],
       "lastStatus": "served"
    }
    ```

6. Client goes back online.

This way you can have all logs as well as a `lastStatus` field for the most current status.

**Comment by nubpro** [comment]
@wei We are actually doing something similar to this, but because of the problem stated in #6363, we had to go for another route.
Instead, we have a columm for each status in the model.

```
type Order @model @auth(rules: [{ allow: owner }]) {
  id: ID!
  acceptedTS: AWSDateTime,
  cookedTS: AWSDateTime,
  servedTS: AWSDateTime,
  lastStatus: String
}
```

Doing so we can circumvent the problem in #6363, as the default conflict resolver, **AUTO_MERGE** will populate any null fields even if there's a conflict in `_version`.

**Comment by wei** [comment]
Thanks for sharing your solution! 

You could also try using the one-to-many relationship models in datastore to keep track of the timestamp and events which should get you around the issue in #6363. 

I guess this issue can be closed or is there any further actions to be taken here? 

We can keep track of the slow network conflict resolver issue in #6363.

**Comment by nubpro** [comment]
> Thanks for sharing your solution!
> 
> You could also try using the one-to-many relationship models in datastore to keep track of the timestamp and events which should get you around the issue in #6363.
> 
> I guess this issue can be closed or is there any further actions to be taken here? We can keep track of the slow network conflict resolver issue in #6363.

We did tried creating a 1:M model called Order Transitions. It ended up creating another issue for us where there's a discrepencies in the two models.
The problem was the lastStatus in the order model is "served". However, the mutation to insert the timestamp of "served" is not delivered to the server yet. Hence, we completely ditched the effort going this route.

Not to mention, we were seeing DataStore errors when inserting multiple transition records. DataStore was sending duplicate mutations to AppSync with the same ID and payload which causes ID conflicts. I believe this is rectified with this PR #6828.

I guess there isn't a need for this feature after all. Thanks!
